packages/integration-karma/test/shadow-dom/ShadowRoot.elementsFromPoint/index.spec.js

```diff
@@ -0,0 +1,131 @@
+    // immediate shadow root, whereas Chrome and Safari return elements above (outside) of that root.
+    // https://crbug.com/1207863#c4
+    let onlyIncludesElementsInImmediateShadowRoot = false;
+    beforeAll(() => {
+        if (process.env.NATIVE_SHADOW) {
+            // detect the Firefox behavior
+            const div = document.createElement('div');
+            div.attachShadow({ mode: 'open' }).innerHTML = '<div>foo</div>';
+            document.body.appendChild(div);
+            const { left, top } = div.shadowRoot.querySelector('div').getBoundingClientRect();
+            const elements = div.shadowRoot.elementsFromPoint(left, top);
+            document.body.removeChild(div);
+            onlyIncludesElementsInImmediateShadowRoot = elements.length === 1;
+        }
+    });
+
+    function test(element, expectedElements) {
+        const { left, top, width, height } = element.getBoundingClientRect();
+        const rootNode = element.getRootNode();
+        const elementsFromPoint = rootNode.elementsFromPoint(left + width / 2, top + height / 2);
+
+        if (onlyIncludesElementsInImmediateShadowRoot) {
+            const mainRootNode = element.getRootNode();
+            expectedElements = expectedElements.filter((el) => el.getRootNode() === mainRootNode);
+        }
+
+        expect(elementsFromPoint).toEqual(expectedElements);
+    }
+
+    // Some browsers have unpredictable behavior for some of these tests
+    const isIE = /(MSIE|Trident)/.test(navigator.userAgent);
+    const isSafari = /Safari/.test(navigator.userAgent) && !/Chrom/.test(navigator.userAgent);
+    const isOldChrome = /Chrom/.test(navigator.userAgent) && process.env.COMPAT;
+    const isOldSafari = isSafari && process.env.COMPAT;
+
+    if (!isOldSafari) {
+        // Safari <v11 doesn't support elementsFromPoint
+        it('non-shadow example', () => {
+            const div = document.createElement('div');
+            div.textContent = 'foo';
+            document.body.appendChild(div);
+
+            const { left, top, width, height } = div.getBoundingClientRect();
+            const elementsFromPoint = document.elementsFromPoint(
+                left + width / 2,
+                top + height / 2
+            );
+
+            expect(elementsFromPoint).toEqual([div, document.body, document.documentElement]);
+        });
+    }
+
+    if (!isIE && !isOldSafari) {
+        // IE returns unpredictable results
+        it('basic shadow example', () => {
+            const elm = createElement('x-container', { is: Container });
+            document.body.appendChild(elm);
+
+            const { body } = document;
+            const html = document.documentElement;
+            const h1 = elm.shadowRoot.querySelector('h1');
+            const inContainer = elm.shadowRoot.querySelector('.in-container');
+            const slottable = elm.shadowRoot.querySelector('x-slottable');
+            const slotted = elm.shadowRoot.querySelector('.slotted');
```

Comment:
I saw that, but I preferred just doing it manually, since with the helper I would have to add `data-id` to every node. But I can change this if you think it's messy.

Reply:
n/m I can see the benefit now. Yes it is less messy this way.